presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,437 @@
+     * @param speechMap the map where keys are offsets into the countdown and values are instructions that should be
+     * spoken to the user.
+     */
+    fun registerSpeechesOnCountdown(duration: Duration, countdown: LiveData<Long>, speechMap: Map<String, String>) {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("register speeches on countdown called,\nduration: $duration\ncountdown: $countdown\n" +
+                    "speechMap: $speechMap")
+        }
+
+        textToSpeech = TextToSpeech(this, this)
+        val canonicalSpeechMap = getCanonicalSpeechMap(speechMap, duration).toMutableMap()
+        val currentCount = countdown.value
+        if (currentCount != null) {
+            // filter out all the speeches that shouldn't be spoken right now, and sort the ones that should in
+            // increasing order of offset.
+            val skippedSpeeches = canonicalSpeechMap.filter { entry ->
+                val currentOffset = duration.seconds - currentCount
+                entry.key < currentOffset && currentOffset - entry.key < speechGracePeriod
+            }.toSortedMap()
+
+            for (speech in skippedSpeeches) {
+                if (!speech.value.second) {
+                    if (LOGGER.isDebugEnabled) {
+                        LOGGER.debug(
+                                "Spoken Instruction: {}, should have been spoken before the service was initialized," +
+                                        " and was still within the grace period.")
+                    }
+                    speakText(speech.value.first)
+                    canonicalSpeechMap[speech.key] = speech.value.copy(second = true)
+                }
+            }
+        }
+
+        futureSpeechData = FutureSpeechData(canonicalSpeechMap, duration, countdown)
+        futureSpeechData?.countdown?.observeForever(::onCountdownChanged)
+    }
+
+    /**
+     * Stops speaking and removes all registered speeches.
+     */
+    fun clear() {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("clear() called")
+        }
+
+        val tts = textToSpeech
+        // Stop speaking and shut down the current text to speech.
+        if (tts != null) {
+            if (tts.isSpeaking) {
+                tts.stop()
+            }
+
+            tts.shutdown()
+            textToSpeech = null
+        }
+
+        futureSpeechData?.countdown?.removeObserver(::onCountdownChanged)
+        futureSpeechData = null
+    }
+
+    /**
+     * Adds the given text to the TextToSpeech's output in a manner consistent with the current queueingBehavior
+     * @param text the text to speak to the user.
+     */
+    fun speakText(text: String) {
+        if (LOGGER.isDebugEnabled) {
+            LOGGER.debug("speakText() called with text \"$text\"")
+        }
+
+        val tts = textToSpeech
+        // Setting this will guarantee the text gets spoken in the case that tts isn't set up yet
+        textToSpeakOnInit = text
```

Comment:
same goes for this except you would also include the else log like this

textToSpeech?.let {
   // version check
} ?: run {
    LOGGER.warn("...")
}